### PR TITLE
KEYCLOAK-6073 - Support different URLs for front and back channel requests in adapters

### PR DESCRIPTION
**Pull request by lgraf**
[KEYCLOAK-6073](https://issues.jboss.org/browse/KEYCLOAK-6073)

[Documentation](https://github.com/keycloak/keycloak-documentation/pull/683)

**Description**
If `auth-server-backchannel-url` is configured the current implementation always resolve backchannel urls during KeycloakDeployment creation (they should never change).

**Absolute `auth-server-url` case**
* During the creation of the KeycloakDeployment frontchannel and backchannel urls are resolved.

**Relative `auth-server-url` case**
* During the creation of the KeycloakDeployment only backchannel urls are resolved. Frontchannel urls will be resolved for each request (AdapterDeploymentContext).

The PR currently contains two commits (implementation, integration tests), maybe this might be helpful during the review. If you prefer a single commit for the review, just let me know and i will squash it.

### CODE REVIEW

File: authz/client/src/main/java/org/keycloak/authorization/client/ServerBackChannelUrlConfiguration.java
Comment (lgraf) [inline authz/client/src/main/java/org/keycloak/authorization/client/ServerBackChannelUrlConfiguration.java]: I am a bit unsure here. Are there reasons not to override the discovered urls with hard coded values?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/BackChannelServletAdapterTest.java
Comment (lgraf) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/BackChannelServletAdapterTest.java]: Most of the changes are in adapter-core, not sure if it necessary to execute the tests against most of the app servers?

**Comment by lukenjmcd** [comment]
We have a similar requirement but in our case we need to support multiple front-end channels.  

We are deploying Keycloak within Kubernetes clusters.  Clients can access this cluster via multiple addresses; some customers access a cluster via a load balancer but others will access it via the the IP or DNS of Kubernetes nodes.

We need the Keycloak Adapter to be able to accept any of these addresses as the token issuer.

We were working on a similar fix to what you've done before we were aware of this PR.  Is it possible to extend your solution to support multiple front-end channels?

**Comment by lgraf** [comment]
Hi @lukenjmcd,
i replied to your comment in the related JIRA [issue](https://issues.jboss.org/browse/KEYCLOAK-6073).

**Comment by mjattiot** [comment]
@lgraf any idea when this PR will be merged and which keycloak release will include it ?
Thank you for your work !

**Comment by mposolda** [comment]
@lgraf @mjattiot Not yet sure. Had longer queue of PRs to review. Hope to go into that in next few weeks.

**Comment by lgraf** [comment]
@mposolda Thanks for the update. Let me know when you find some time for the review, so that i can rebase the branch on master before.

**Comment by viliampucik** [comment]
Please, any plans to move this upstream? This feature would be a life saver.

**Comment by stianst** [comment]
For everyone interested in this PR it would be great if you could review/test it, as that would help us to make sure this is working properly.

**Comment by stianst** [comment]
@mposolda can you get someone from your team to review this one please?\

**Comment by lgraf** [comment]
Rebased on current master.

**Comment by stianst** [comment]
@mposolda can you get someone from your team to review this one please?

**Comment by mposolda** [comment]
@stianst Funny, this is something, which I added few years ago, then we removed it, and now we have contribution to add it back :)

I've quickly looked at the PR. Adding @pskopek as reviewer to do more proper reviews.

@lgraf Few question/concerns:
- If people doesn't configure new option "auth-server-backchannel-url", the backchannel urls will be same as frontchannel url right? In other words, it is backwards compatible and existing deployments won't be affected, is it correct?
- Documentation PR will be also needed. I am adding flag "Missing documentation" . In the "Securing application" guide, there is chapter, which specifies all the adapter options. Not sure if it makes sense to document it somewhere else. If there are any specific use-cases solved by this, it may be benefitical to document in particular chapter (EG. reverse proxy chapter or adapter clustering chapter)
- The PR is missing updating of Wildfly/EAP subsystems. The new option needs to be added to them as well. See for example this commit for some details and changes in Wildfly/as7/wf-8 subsystems: https://source.jboss.org/changelog/Keycloak?cs=3777dc45d0a522f458bd7ee9dd31c6d3d91e8fd1

**Comment by stianst** [comment]
@mposolda I know - it didn't make sense in the past when we didn't have the hostname SPI, but now it does ;)

**Comment by lgraf** [comment]
> @lgraf Few question/concerns:
> 
> * If people doesn't configure new option "auth-server-backchannel-url", the backchannel urls will be same as frontchannel url right? In other words, it is backwards compatible and existing deployments won't be affected, is it correct?

Yes, if no backchannel URL is configured the new option will be ignored by the adapter and all URL`s will be resolved by using the frontchannel URL ("auth-server-url"). In turn the changes should be backwards compatible, existing deployments should be upgradeable without affecting the behaviour of the deployment.

As far i can tell, there are also existing adapter [tests](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/DemoServletsAdapterTest.java), which should reveal potential backwards compatibility issues.

> * Documentation PR will be also needed. I am adding flag "Missing documentation" . In the "Securing application" guide, there is chapter, which specifies all the adapter options. Not sure if it makes sense to document it somewhere else. If there are any specific use-cases solved by this, it may be benefitical to document in particular chapter (EG. reverse proxy chapter or adapter clustering chapter)

A documentation PR for the adapter option is already exists and contains a reference to the hostname SPI documentation (link in PR description). I rebased this PR on the current documentation master as well.

One thing that may be worth noting in the [relative uri optimization](https://www.keycloak.org/docs/latest/securing_apps/index.html#relative-uri-optimization) section is, that only the frontchannel URI's will be resolved from the current request URL, when a backchannel-url is used. Not sure if this obivisious?

The use-case this feature solves is mostly application/deployment specific. I am not sure if it necessary to document it in other places as well. But i would be happy to get other opinions on this topic.

> * The PR is missing updating of Wildfly/EAP subsystems. The new option needs to be added to them as well. See for example this commit for some details and changes in Wildfly/as7/wf-8 subsystems: https://source.jboss.org/changelog/Keycloak?cs=3777dc45d0a522f458bd7ee9dd31c6d3d91e8fd1

This is necessary to make the new adapter option available for "external" (configuration not included inside the .war) Wildfly/EAP adapter configuration? I will try to find some time next week to add the required changes to the PR. Thanks for pointing this out.

> * Do you have a chance to rebase again? Test failures were fixed in master in the meantime. Thanls

Done

**Comment by mposolda** [comment]
> > @lgraf Few question/concerns:
> > 
> > * If people doesn't configure new option "auth-server-backchannel-url", the backchannel urls will be same as frontchannel url right? In other words, it is backwards compatible and existing deployments won't be affected, is it correct?
> 
> Yes, if no backchannel URL is configured the new option will be ignored by the adapter and all URL`s will be resolved by using the frontchannel URL ("auth-server-url"). In turn the changes should be backwards compatible, existing deployments should be upgradeable without affecting the behaviour of the deployment.
> 
> As far i can tell, there are also existing adapter [tests](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/DemoServletsAdapterTest.java), which should reveal potential backwards compatibility issues.

That's what I though. Thanks for the clarification.

> 
> > * Documentation PR will be also needed. I am adding flag "Missing documentation" . In the "Securing application" guide, there is chapter, which specifies all the adapter options. Not sure if it makes sense to document it somewhere else. If there are any specific use-cases solved by this, it may be benefitical to document in particular chapter (EG. reverse proxy chapter or adapter clustering chapter)
> 
> A documentation PR for the adapter option is already exists and contains a reference to the hostname SPI documentation (link in PR description). I rebased this PR on the current documentation master as well.
> 
> One thing that may be worth noting in the [relative uri optimization](https://www.keycloak.org/docs/latest/securing_apps/index.html#relative-uri-optimization) section is, that only the frontchannel URI's will be resolved from the current request URL, when a backchannel-url is used. Not sure if this obivisious?
> 
> The use-case this feature solves is mostly application/deployment specific. I am not sure if it necessary to document it in other places as well. But i would be happy to get other opinions on this topic.


Thanks, agree that sentence/note in relative URI section will be good. Added the note to your documentation PR about this as well.

> 
> > * The PR is missing updating of Wildfly/EAP subsystems. The new option needs to be added to them as well. See for example this commit for some details and changes in Wildfly/as7/wf-8 subsystems: https://source.jboss.org/changelog/Keycloak?cs=3777dc45d0a522f458bd7ee9dd31c6d3d91e8fd1
> 
> This is necessary to make the new adapter option available for "external" (configuration not included inside the .war) Wildfly/EAP adapter configuration? I will try to find some time next week to add the required changes to the PR. Thanks for pointing this out.

Yes, exactly. This is for external Wildfly/EAP adapter configuration in standalone.xml file. Thanks.

> 
> > * Do you have a chance to rebase again? Test failures were fixed in master in the meantime. Thanls
> 
> Done

Thanks, travis test failure is gone now.

**Review by andymunro — APPROVED** [review]

**Comment by stianst** [comment]
@mposolda I wonder if a better design would be to have the adapter load URLs from well-known/openid-configuration, then the URLs can be controlled from the Keycloak server side. That would be required by any third party library. Basically openid-configuration would return the configured URL for browser facing URLs, while showing URLs based on the request URI for backchannel requests. Should also perhaps improve on the hostname provider so it perhaps can have one configured internal and another external URL. Will write this up in more details on the dev ML, but just wanted to give a heads up before this is merged.

**Review by stianst — COMMENTED** [review]
Please hold off on this PR until we've discussed a potentially better way to address this

**Comment by stianst** [comment]
See https://github.com/keycloak/keycloak-community/blob/master/design/hostname-default-provider.md for a proposal of a solution that would make it easier to configure Keycloak as well as add support for different front/back channel URLs both for Keycloak adapters and third-party libraries.

**Comment by stianst** [comment]
PR for new hostname provider is ready here: https://github.com/keycloak/keycloak/pull/6388

**Comment by stianst** [comment]
Thanks for the effort in this PR, but I do believe the new approach with the default hostname provider being able to set correct URLs in discovery endpoint is a better approach and also would support different URLs in third party libraries. For Keycloak adapters we should update to load issuer and endpoints from discovery endpoint instead of hard coding.

**Comment by lgraf** [comment]
@stianst I agree that new approach is the better suited to solve this requirement. This PR can be closed.

**Comment by stianst** [comment]
@lgraf Thanks, and thanks again for your contribution and I'm sorry for the inconvenience this may have caused you.

**Comment by stianst** [comment]
Closing this as we have decided to go with loading the endpoints dynamically from the Keycloak server using the OIDC Discovery endpoint now that Keycloak server supports advertising different URLs for frontend and backend requests.

**Comment by stianst** [comment]
FIY
New default provider PR here: https://github.com/keycloak/keycloak/pull/6388
Adapters loading URLs from discovery endpoint PR here: https://github.com/keycloak/keycloak/pull/6412

**Review by andymunro — COMMENTED** [review]
Looks good.

**Comment by algakzru** [comment]
I am a little bit confused with different URLs for frontend and backend...

In my scenario I have a Kubernetes cluster with Keycloak, Vuejs frontend and Spring-Boot backend.

Spring-Boot Backend client is located on the same network as the Keycloak server, so it can communicate directly with Keycloak using an internal domain name.
This is how I set Keycloak url for Spring-Boot adapter :
```
keycloak:
  auth-server-url: http://keycloak.default.svc.cluster.local:8080/auth
```

Vuejs Frontend app is connecting to Keycloak from browser, so it can not communicate with Keycloak using an internal domain name, instead it must use external domain name.
This is how I set Keycloak url for Javascript adapter :
```
{
    "auth-server-url": "https://keycloak.some-external-domain.com/auth"
}
```

But currently it doesn't work with different urls, it only works when both frontend and backend use the same url to connect to Keycloak, in my case "https://keycloak.some-external-domain.com/auth".

I have read https://github.com/keycloak/keycloak-community/blob/master/design/hostname-default-provider.md carefully, but still unclear whether it solves my problem or not.

If yes, how to do it ?

I tried to add `KEYCLOAK_FRONTEND_URL` to Keycloak deployment, but after that Vuejs Frontend is not able to connect to Keycloak
```
spec:
  template:
    spec:
      containers:
      - env:
        - name: KEYCLOAK_FRONTEND_URL
          value: “https://keycloak.some-external-domain.com”
```

**Comment by stianst** [comment]
KEYCLOAK_FRONTEND_URL needs to include the context-path Keycloak is exposed on, so you probably only need to change it to "https://keycloak.some-external-domain.com/auth".

**Comment by algakzru** [comment]
@stianst 
Thank you, now it finally works.

I was following the example in the documentation https://github.com/keycloak/keycloak-community/blob/master/design/hostname-default-provider.md where `auth-server-url` in the adapter is set to `https://keycloak.mycompany.local/auth`, but the `url` of the default hostname provider is set to `https://auth.mycompany.com`, without the context-path.
That is what confused me :)

**Comment by algakzru** [comment]
@stianst

**Pull request closed without merging**
